File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/validator/common/update/AbstractTooCloseUpdateValidator.java
package cz.fb.manaus.reactor.betting.validator.common.update;

import com.google.common.base.Preconditions;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.reactor.betting.BetContext;
import cz.fb.manaus.reactor.betting.validator.ValidationResult;
import cz.fb.manaus.reactor.betting.validator.Validator;
import cz.fb.manaus.reactor.price.PriceService;
import cz.fb.manaus.reactor.rounding.RoundingService;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.OptionalDouble;
import java.util.Set;

abstract public class AbstractTooCloseUpdateValidator implements Validator {
    private final Set<Integer> closeSteps;
    @Autowired
    private RoundingService roundingService;
    @Autowired
    private PriceService priceService;

    public AbstractTooCloseUpdateValidator(Set<Integer> closeSteps) {
        this.closeSteps = closeSteps;
    }

    @Override
    public boolean isUpdateOnly() {
        return true;
    }

    @Override
    public ValidationResult validate(BetContext context) {
        double oldOne = context.getOldBet().get().getRequestedPrice().getPrice();
        double newOne = context.getNewPrice().get().getPrice();
        if (Price.priceEq(newOne, oldOne)) return ValidationResult.REJECT;
        for (Integer step : closeSteps) {
            Preconditions.checkArgument(step != 0);
            OptionalDouble closePrice;
            if (step > 0) {
                closePrice = roundingService.increment(oldOne, step);
            } else {
                closePrice = roundingService.decrement(oldOne, -step);
            }
            if (closePrice.isPresent() && Price.priceEq(newOne, closePrice.getAsDouble()))
                return ValidationResult.REJECT;
        }
        return ValidationResult.ACCEPT;
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/validator/common/update/AbstractTooCloseUpdateValidator.kt
package cz.fb.manaus.reactor.betting.validator.common.update

import cz.fb.manaus.core.model.priceEq
import cz.fb.manaus.reactor.betting.BetEvent
import cz.fb.manaus.reactor.betting.validator.UpdateOnlyValidator
import cz.fb.manaus.reactor.betting.validator.ValidationResult
import cz.fb.manaus.reactor.rounding.RoundingService
import org.springframework.beans.factory.annotation.Autowired

abstract class AbstractTooCloseUpdateValidator(private val closeSteps: Set<Int>) : UpdateOnlyValidator {
    @Autowired
    private lateinit var roundingService: RoundingService

    override fun validate(event: BetEvent): ValidationResult {
        val oldOne = event.oldBet!!.requestedPrice.price
        val newOne = event.newPrice!!.price
        if (newOne priceEq oldOne) return ValidationResult.NOP
        val minPrice = event.account.provider.minPrice
        val tagPredicate = event.account.provider::matches
        val containsEqualPrice = closeSteps
                .onEach { require(it != 0) }
                .mapNotNull {
                    when {
                        it > 0 -> roundingService.increment(oldOne, it, tagPredicate)
                        else -> roundingService.decrement(oldOne, -it, minPrice, tagPredicate)
                    }
                }.any { newOne priceEq it }
        return if (containsEqualPrice) ValidationResult.NOP else ValidationResult.ACCEPT
    }

}
